### ISSUE TITLE
mkdir(a); mkdir(a) logs an unhandled error in SFTP

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3033]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#3033|
|Type|defect|
|Created|2008-02-14 01:19:30Z|
|Branch|https://github.com/twisted/twisted/tree/sftp-eexist-3033|

mkdir(a); mkdir(a) logs an unhandled error in SFTP.

The bug is in _ebStatus in filetransfer.py. At the moment, it does not handle EEXIST. Instead it should send an error message with the appropriate error codes and _not_ log an error.

For example:

```
=== modified file 'twisted/conch/ssh/filetransfer.py'
--- twisted/conch/ssh/filetransfer.py   2006-06-13 12:54:33 +0000
+++ twisted/conch/ssh/filetransfer.py   2008-02-14 01:04:55 +0000
@@ -421,6 +421,17 @@
             elif reason.value.errno == errno.EACCES: # permission denied
                 code = FX_PERMISSION_DENIED
                 message = reason.value.strerror
+            elif reason.value.errno == errno.EEXIST:
+                # XXX: JonathanLange 2008-02-14:
+                # We'll set the code, but leave the message unchanged. The
+                # message is important for tools that do string-based matching
+                # on errors. In particular, one common way of checking for
+                # 'the directory that you are trying to make already exists'
+                # is to check that the error message is 'mkdir failed'. This
+                # is the tragedy of SFTP. See
+                # bzrlib.transport.sftp.SFTPTransport._translate_io_exception
+                # for an example.
+                code = FX_FILE_ALREADY_EXISTS
             else:
                 log.err(reason)
         elif reason.type == EOFError: # EOF
```

I would submit a proper patch, but writing unit tests for Conch is too hard.

<details><summary>Searchable metadata</summary>

```
trac-id__3033 3033
type__defect defect
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_sftp_eexist_3033 branches/sftp-eexist-3033
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1202951970000000 1202951970000000
changetime__1211888934000000 1211888934000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @therve|
|-|-|

Someone has to add tests anyway.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [23171]) Branching to 'sftp-eexist-3033'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [23172]) Commit the fix with a test.

Refs [#3033](https://github.com/twisted/twisted/issues/3033)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [23173]) Fix error message.

Refs [#3033](https://github.com/twisted/twisted/issues/3033)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to z3p|
|-|-|

Please review. I tried to take into account what jml said about the message, but with adding a little bit more information.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

Unfortunately, the message has to be 'mkdir failed' exactly.

```python
        # paramiko seems to generate detailless errors.
        self._translate_error(e, path, raise_generic=False)
        if getattr(e, 'args', None) is not None:
            if (e.args == ('No such file or directory',) or
                e.args == ('No such file',)):
                raise NoSuchFile(path, str(e) + more_info)
            if (e.args == ('mkdir failed',)):
                raise FileExists(path, str(e) + more_info)
            # strange but true, for the paramiko server.
            if (e.args == ('Failure',)):
                raise failure_exc(path, str(e) + more_info)
            mutter('Raising exception with args %s', e.args)
        if getattr(e, 'errno', None) is not None:
            mutter('Raising exception with errno %s', e.errno)
        raise e
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

jml, can you elaborate on this?  For example, how does the requirement for the failure message being equal to "mkdir failed" relate to this pure-OpenSSH sftp transcript?

```
exarkun@charm:~$ sftp home.intarweb.us
Connecting to home.intarweb.us...
sftp> mkdir foobar
sftp> mkdir foobar
Couldn't create directory: Failure
sftp> mkdir barbaz
sftp> rename barbaz foobar
Couldn't rename file "/home/exarkun/barbaz" to "/home/exarkun/foobar": Failure
sftp> ^D
exarkun@charm:~$
```

It seems paramiko is overly specific in its checking and should be fixed if it can't recognize `FX_FILE_ALREADY_EXISTS`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

The code changes look good to me, and I think the more expressive error message conch is now trying to generate is better than just the text paramiko wants.  However, I can't seem to observe the change.  I'm not sure I even completely understand what the bug is supposed to be.  Does the Conch server generate the wrong error text to send back to the client?  Does the Conch client fail to properly interpret error messages (error codes?) from the server?

The original bug report should probably have been more detailed.  Since therve made some changes, though, I guess he understood it, so I'll assign back to him for clarification.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [23421]) Restore back error message

Refs [#3033](https://github.com/twisted/twisted/issues/3033)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

The ticket is pretty clear to me: the sftp server logged an unhandled error (with log.err). The goal is just to recognize EEXIST to prevent this useless logging.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_13"></a>|@jml set owner to @jml|
|-|-|

I'll investigate and clarify what Paramiko, OpenSSH and the standard require here, then review the patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

I think it'd be fine if the failure had more information in it.  However, the branch is still a good improvement over trunk.  We can file a separate enhancement for improving the user-facing error reporting of cftp later, if we care.  Regarding the _exact_ text of the message, since bzrlib uses paramiko, not conch, in the future I don't think we need to consider the message that bzrlib expects.

I think another potential improvement here would be to actually define `FX_FILE_ALREADY_EXISTS` distinct from `FX_FAILURE` (version 6 of the protocol gives it a value of 11).  It would be nice if conch had implementations of the widely used versions of sftp though, rather than just the latest, or just a random version, or a random mix of different versions.  3 and 4 seem to be the most widely used, and 6 is the most recent.  However, that's all for some other ticket as well.
(Actually, I suppose it's pretty related, since presumably the reason the string must be exactly "mkdir failed" is that Conch sends back `FX_FAILURE` instead of `FX_FILE_ALREADY_EXISTS`?)

Okay, so then, it seems like maybe we should define the correct value for `FX_FILE_ALREADY_EXISTS` now.  It has been defined for almost six years (it seems to have been introduced in version 4 of the protocol).  This presumably would remove the need for the exact string messages, but that can still be a separate issue.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve removed owner|
|-|-|

I've changed the value of FX_FILE_ALREADY_EXISTS. I had a little problem in vfs where a test was actually testing the wrong thing because of a problem in the inmem implementation.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_16"></a>|@jml commented|
|-|-|

FWIW, Paramiko and OpenSSH both implement only version 3 of the protocol, which is defined by [IETF draft 02](http://tools.ietf.org/wg/secsh/draft-ietf-secsh-filexfer/draft-ietf-secsh-filexfer-02.txt). This protocol version doesn't include `FX_FILE_ALREADY_EXISTS`.

I think that mixing protocol versions is something that Paramiko does as well. I haven't yet looked at the OpenSSH code, but I'd be surprised if it implemented v3 as strictly defined by the IETF draft.

Still, if OpenSSH and Paramiko behave reasonably, then I am +1 on using FX_FILE_ALREADY_EXISTS = 11 as the error code and +1 on attempting to resolve the "exact string" issue in another ticket.

See also http://www.greenend.org.uk/rjk/2007/sftpversions.html.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @therve|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set status to `closed`|
|-|-|

(In [23710]) Merge sftp-eexist-3033

Author: therve, jml
Reviewers: exarkun, jml
Fixes [#3033](https://github.com/twisted/twisted/issues/3033)

Manage EEXIST in conch.ssh.filetransfer, to return FX_FILE_ALREADY_EXISTS
instead of a generic error and prevent a logged error. It also changes the
value of FX_FILE_ALREADY_EXISTS to update it to the value introduced by
version 4 of the protocol.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
